usr/lib/soft_stdll/soft_specific.c

```diff
@@ -3051,3 +3075,540 @@ CK_RV token_specific_aes_cmac(STDLL_TokData_t *tokdata, CK_BYTE *message,
+                                         TEMPLATE *publ_tmpl,
+                                         TEMPLATE *priv_tmpl)
+{
+
+    CK_ATTRIBUTE *attr = NULL, *ec_point_attr, *value_attr, *parms_attr;
+    EC_KEY *ec_key = NULL;
+    BN_CTX *ctx = NULL;
+    CK_BYTE *ecpoint = NULL, *enc_ecpoint = NULL,*d = NULL;
+    CK_ULONG ecpoint_len, enc_ecpoint_len, d_len;
+    CK_RV rc;
+
+    UNUSED(tokdata);
+
+    if (!template_attribute_find(publ_tmpl, CKA_ECDSA_PARAMS, &attr)) {
+        TRACE_ERROR("Could not find CKA_ECDSA_PARAMS in the template\n");
+        rc = CKR_TEMPLATE_INCOMPLETE;
+        goto out;
+    }
+
+    rc = make_ec_key_from_params(attr->pValue, attr->ulValueLen, &ec_key);
+    if (rc != CKR_OK)
+        goto out;
+
+    if (!EC_KEY_generate_key(ec_key)) {
+        TRACE_ERROR("Failed to generate an EC key.\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    ctx = BN_CTX_new();
+    if (ctx == NULL) {
+        rc = CKR_HOST_MEMORY;
+        goto out;
+    }
+
+    ecpoint_len = EC_KEY_key2buf(ec_key, POINT_CONVERSION_UNCOMPRESSED,
+                                 &ecpoint, ctx);
+    if (ecpoint_len == 0) {
+        TRACE_ERROR("Failed to get the EC Point compressed.\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    rc = ber_encode_OCTET_STRING(FALSE, &enc_ecpoint, &enc_ecpoint_len,
+                                 ecpoint, ecpoint_len);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_encode_OCTET_STRING failed\n");
+        goto out;
+    }
+
+    rc = build_attribute(CKA_EC_POINT, enc_ecpoint, enc_ecpoint_len,
+                         &ec_point_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("build_attribute for CKA_EC_POINT failed rc=0x%lx\n", rc);
+        goto out;
+    }
+    template_update_attribute(publ_tmpl, ec_point_attr);
+
+    d_len = EC_KEY_priv2buf(ec_key, &d);
+    if (d_len == 0) {
+        TRACE_ERROR("Failed to get the EC private key.\n");
+        rc = CKR_FUNCTION_FAILED;
+        goto out;
+    }
+
+    rc = build_attribute(CKA_VALUE, d, d_len, &value_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("build_attribute for CKA_VALUE failed, rc=0x%lx\n", rc);
+        goto out;
+    }
+    template_update_attribute(priv_tmpl, value_attr);
+
+    /* Add CKA_ECDSA_PARAMS to private template also */
+    rc = build_attribute(CKA_ECDSA_PARAMS, attr->pValue, attr->ulValueLen,
+                         &parms_attr);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("build_attribute for CKA_ECDSA_PARAMS failed, rc=0x%lx\n",
+                     rc);
+        goto out;
+    }
+    template_update_attribute(priv_tmpl, parms_attr);
+
+    rc = CKR_OK;
+
+out:
+    if (ctx)
+        BN_CTX_free(ctx);
+    if (ec_key != NULL)
+        EC_KEY_free(ec_key);
+    if (ecpoint != NULL)
+        OPENSSL_free(ecpoint);
+    if (enc_ecpoint != NULL)
+        free(enc_ecpoint);
+    if (d != NULL)
+        OPENSSL_free(d);
```

Comment:
Same for the other openssl free routines, they all check if the arg is null. For free() we have to keep the check.